pkg/images/pull.go

```diff
@@ -121,41 +122,66 @@ func handleTarObject(ctx context.Context, tr *tar.Reader, hdr *tar.Header, conf
 		}
 	case tar.TypeReg:
 		compressed := strings.HasSuffix(dstPath, ".zst")
-		if compressed {
-			image = strings.TrimSuffix(dstPath, ".zst")
-			if _, err := os.Stat(image); err == nil {
-				return image, os.ErrExist
-			}
-
-			cmd := exec.CommandContext(ctx, "zstd", "-d", "-", "-o", image)
-			cmd.Stdin = tr
 
-			if _, err := cmd.Output(); err != nil {
-				var e *exec.ExitError
-				if errors.As(err, &e) {
-					fmt.Fprintf(os.Stderr, string(e.Stderr))
-				}
-				return image, fmt.Errorf("failed during zst decompression of %s: %w", hdr.Name, err)
-			}
+		// Copy the target file out to the host (compressed or not).
+		// On failure, clean up the temporary file; otherwise move it
+		// to the target destination.
+		tmpFile, err := os.CreateTemp("", filepath.Base(hdr.Name))
```

Comment:
This seems more  complicated that it has to be.
Why not just do:

```
// copy file to dstPath
if compressed {
    // uncompress
    if !cache {
            // rm dstPath
    }
}

```

Reply:
I figured that copying to a potentially RAM-backed tmpfs would be a little more optimal, perhaps skip the extra main filesystem write if caching isn't needed. But you're probably right that this is premature optimization.

Reply:
Also has the property of not putting a half-baked file on the filesystem there in case of error/truncation. Bit of a corner case.

Reply:
I'm probably overthinking the problem though, happy to fix it up.

Reply:
What about https://go.dev/pkg/io/?m=old#TeeReader for an in-memory copy? Saves you all the dealing with temp files.

Reply:
Tee is a bad idea because the file is massive.